### PR TITLE
feat(data-structures/unstable): rework IndexedHeap with set() upsert and generic priorities

### PR DESCRIPTION
**Pull request by tomas-zijdemans**
Rework IndexedHeap around Map-style upsert and generic priorities.

- Replace `update()` and `pushOrUpdate()` with a single `set()` upsert,
  matching Map's API and the natural shape for Dijkstra-style relaxation.
- Generalize priorities beyond `number` via a new `P` type parameter,
  enabling max-heaps (`descend`), `bigint`, and tuple priorities for stable
  tie-breaking. `P = number` remains the default.
- Add constructor loading + `compare` option, and `IndexedHeap.from(other)`
  inheritance of the source comparator.
- The `compare` option is required at the type level when `P` is not a
  primitive, to prevent silent mis-ordering with the default `ascend`.
- Expand docs and tests around the grouped API changes.

The constructor takes `compare` via an options bag (to leave room for future options) and defaults to `ascend` (min-heap) because the primary use cases (Dijkstra-style relaxation, task schedulers) are min-heap.

### CODE REVIEW

**Review by fibibot — APPROVED** [review]
The `update`/`pushOrUpdate` collapse into a single `set()` upsert matches Map's API and is the natural shape for Dijkstra-style relaxation. Generic priority `P = number` with the `compare` option keeps the default ergonomic for the common case and unlocks tuple/bigint/descend variants without internal forks.

`from(otherHeap)` correctly inherits `#compare` and reuses the source's heap-array order (still valid under the same comparator); the override branch re-heapifies via Floyd's siftDown in O(n). `checkPriority` narrowing to `typeof === "number" && isNaN` is right — `NaN` only exists in `number`.

Tests cover the new surface well: tuple stable tie-breaking, bigint past `MAX_SAFE_INTEGER`, `set` no-op on identical priority, NaN rejection on both absent and present keys, ArrayLike redirected to `from()`.

- nit: `from(otherHeap, { compare: collection.#compare })` re-heapifies even when the comparator is identical to the source's. Cheap edge case to skip, but function identity is fragile so probably not worth optimising.

**Review by bartlomieju — COMMENTED** [review]
Solid work — generalization is sound, tests cover the new dimensions well, and the docs are thorough. A handful of inline notes below.

One overall comment: the PR title says "align with heap APIs" but the resulting shape still diverges from `BinaryHeap` in two visible ways — `BinaryHeap` takes `compare` positionally and defaults to `descend` (max-heap), while `IndexedHeap` takes `compare` via an options bag and defaults to `ascend` (min-heap). Both choices are defensible (min-heap matches the Dijkstra-style use cases this class is built for), but it's worth calling the divergence out in the PR description so it's a conscious decision rather than an accidental one.

File: data_structures/unstable_indexed_heap.ts [hunk 1]
```diff
@@ -106,13 +126,129 @@ function checkPriority(priority: number): void {
...
+   *   compare: descend,
+   * });
+   * assertEquals(heap.peek(), { key: "b", priority: 5 });
+   * ```
+   */
+  constructor(
+    entries?: Iterable<readonly [K, P]> | null,
+    options?: { compare?: (a: P, b: P) => number },
+  ) {
+    const compare = options?.compare ?? (ascend as (a: P, b: P) => number);
```
Comment (bartlomieju) [inline data_structures/unstable_indexed_heap.ts:258]: `compare` is optional regardless of `P`, but the `ascend` default only works for primitives (`number`, `bigint`, `string`). With `P = [number, number]` and no `compare`, `ascend` silently produces nonsense ordering — no TS error and no runtime error, the heap just isn't a heap.

Options: a typed overload that requires `compare` when `P` isn't a primitive, or at minimum a class-level docs note saying "`compare` is required if `P` is an object type." A runtime probe on the first comparison would catch it eagerly too.

File: data_structures/unstable_indexed_heap.ts [hunk 2]
```diff
@@ -106,13 +126,129 @@ function checkPriority(priority: number): void {
...
+    if (typeof compare !== "function") {
+      throw new TypeError(
+        "Cannot construct an IndexedHeap: the 'compare' option is not a function",
+      );
+    }
+    this.#compare = compare;
+    if (entries === undefined || entries === null) return;
+    if (
+      typeof entries !== "object" && typeof entries !== "string" ||
```
Comment (bartlomieju) [inline data_structures/unstable_indexed_heap.ts]: The `!== "string"` exception lets strings through the iterability check, but a string isn't a valid `Iterable<[K, P]>`. `new IndexedHeap("foo" as any)` will fail inside `#bulkLoad` with a confusing destructuring error instead of the helpful "did you mean `IndexedHeap.from`?" message that this guard exists to give.

Either drop the `"string"` exception or add a test pinning the intended behavior.

File: data_structures/unstable_indexed_heap.ts [hunk 3]
```diff
@@ -557,7 +712,7 @@ export class IndexedHeap<K> implements Iterable<HeapEntry<K>> {
     const oldPriority = entry.priority;
     if (priority === oldPriority) return;
```
Comment (bartlomieju) [inline data_structures/unstable_indexed_heap.ts:803]: With tuple/object priorities this is reference equality, so `heap.set("a", [5, 0])` is **not** a no-op even when the values are equivalent to the existing priority — it replaces the reference and runs a sift. That's the correct/cheap behavior, but it's a footgun worth one line of docs near the class-level "treat object priorities as immutable" note.

File: data_structures/unstable_indexed_heap.ts [hunk 4]
```diff
@@ -744,18 +899,19 @@ export class IndexedHeap<K> implements Iterable<HeapEntry<K>> {
    *
    * @returns An array of entries in arbitrary (heap-internal) order.
    */
-  toArray(): HeapEntry<K>[] {
+  toArray(): HeapEntry<K, P>[] {
```
Comment (bartlomieju) [inline data_structures/unstable_indexed_heap.ts:996]: The `peek()` JSDoc now warns that object priorities share a reference with the heap. `toArray()` and `[Symbol.iterator]` have the same property — they return fresh wrappers but the `priority` field is the same reference held internally. Worth mirroring the warning on both so the caveat isn't easy to miss.

File: data_structures/unstable_indexed_heap.ts [hunk 5]
```diff
@@ -777,7 +933,7 @@ export class IndexedHeap<K> implements Iterable<HeapEntry<K>> {
    *
    * @returns An iterator yielding entries in arbitrary (heap-internal) order.
    */
-  *[Symbol.iterator](): IterableIterator<HeapEntry<K>> {
+  *[Symbol.iterator](): IterableIterator<HeapEntry<K, P>> {
```
Comment (bartlomieju) [inline data_structures/unstable_indexed_heap.ts:1034]: Same priority-reference caveat as `toArray()` above — the yielded `priority` is the heap's live reference for object-typed priorities. Worth one line referencing the class-level note.

File: data_structures/unstable_indexed_heap.ts (hunk 1 above)
Comment (tomas-zijdemans) [reply data_structures/unstable_indexed_heap.ts:258]: Typed conditional overload: new internal IndexedHeapCompareArgs<P> rest-tuple type forces compare at the type level when P isn't number | bigint | string. Applied to both the public constructor and the new iterable/array-like overload of static from(). The IndexedHeap-source overload of static from() keeps compare optional (inherits from source).

File: data_structures/unstable_indexed_heap.ts (hunk 2 above)
Comment (tomas-zijdemans) [reply data_structures/unstable_indexed_heap.ts]: Dropped "string" exception in both iterability guards — strings now hit the helpful "did you mean IndexedHeap.from?" / "not iterable or array-like" TypeError instead of failing inside #bulkLoad.

File: data_structures/unstable_indexed_heap.ts (hunk 3 above)
Comment (tomas-zijdemans) [reply data_structures/unstable_indexed_heap.ts:803]: set() reference-equality docs: extended the class-level priority-immutability note with the === short-circuit behavior; mirrored a more detailed version on set() JSDoc with a cross-link.

File: data_structures/unstable_indexed_heap.ts (hunk 4 above)
Comment (tomas-zijdemans) [reply data_structures/unstable_indexed_heap.ts:996]: Priority-reference pointer: added a one-liner to toArray() and [Symbol.iterator] pointing at the class-level note; shortened peek()'s warning to the same one-liner for consistency. Class-level note is the single source of truth.

**Comment by tomas-zijdemans** [comment]
Sorry @bartlomieju. I realize the original PR title and description was inadequate. I've rephrased them.

I've updated the PR based on your feedback

**Comment by tomas-zijdemans** [comment]
Please note that the CI is flagging a typo: `Warning: "mis" should be "miss" or "mist".`

... but I think this is a false positive "mis-order" is correct, "miss-order" would be wrong

**Review by bartlomieju — APPROVED** [review]
